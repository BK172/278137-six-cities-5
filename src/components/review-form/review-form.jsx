import React, {PureComponent, Fragment} from "react";
import PropTypes from "prop-types";
import {postReview} from "../../store/api-actions";
import {connect} from "react-redux";
import {ReviewFormTextAreaLength, ReviewFormRatings, ResponseType} from "../../constants";

class ReviewForm extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      rating: ``,
      review: ``,
      isFormValid: false,
      isFormWaitingResponse: false,
      reviewPostingError: ``,
    };

    this._handleFormSubmit = this._handleFormSubmit.bind(this);
    this._handleInputChange = this._handleInputChange.bind(this);
    this._handleClearFormFields = this._handleClearFormFields.bind(this);
    this._handleChangeFormWaitingFlag = this._handleChangeFormWaitingFlag.bind(this);
    this._handleChangeReviewPostingError = this._handleChangeReviewPostingError.bind(this);
  }

  componentDidUpdate() {
    const {rating, review} = this.state;

    if (rating && review.length >=
        ReviewFormTextAreaLength.MIN && review.length <= ReviewFormTextAreaLength.MAX
    ) {
      this.setState({isFormValid: true});
    } else {
      this.setState({isFormValid: false});
    }
  }

  _handleFormSubmit(evt) {
    evt.preventDefault();

    const {offerId, postReviewAction} = this.props;
    const review = this.state.review;
    const rating = this.state.rating;
    const onClearFormFields = this._handleClearFormFields;
    const onChangeFormWaitingFlag = this._handleChangeFormWaitingFlag;
    const onChangeReviewPostingError = this._handleChangeReviewPostingError;

    this._handleChangeFormWaitingFlag(true);
    this._handleChangeReviewPostingError(``);
    postReviewAction({
      review,
      rating,
      offerId,
      onClearFormFields,
      onChangeFormWaitingFlag,
      onChangeReviewPostingError,
    });
  }

  _handleInputChange({target}) {
    const {name, value} = target;
    this.setState({[name]: value});
  }

  _handleClearFormFields() {
    this.setState({
      rating: ``,
      review: ``,
    });
  }

  _handleChangeFormWaitingFlag(isWaitingFlag) {
    this.setState(() => ({
      isFormWaitingResponse: isWaitingFlag,
    }));
  }

  _handleChangeReviewPostingError(reviewPostingError) {
    this.setState(() => ({
      reviewPostingError,
    }));
  }

  render() {
    return (
      <form className="reviews__form form" action="#" method="post" onSubmit={this._handleFormSubmit}>
        <label className="reviews__label form__label" htmlFor="review">Your review</label>
        <div className="reviews__rating-form form__rating">
          {ReviewFormRatings.map(({mark, title}) => (
            <Fragment key={title}>
              <input
                className="form__rating-input visually-hidden"
                name="rating"
                value={mark}
                id={title}
                type="radio"
                checked={mark === this.state.rating}
                onChange={this._handleInputChange}
              />
              <label htmlFor={title} className="reviews__rating-label form__rating-label" title={title}>
                <svg className="form__star-image" width={37} height={33}>
                  <use xlinkHref="#icon-star"/>
                </svg>
              </label>
            </Fragment>
          ))}
        </div>
        <textarea
          className="reviews__textarea form__textarea"
          id="review"
          name="review"
          value={this.state.review}
          onChange={this._handleInputChange}
          placeholder="Tell how was your stay, what you like and what can be improved"
        />
        <div className="reviews__button-wrapper">
          <p className="reviews__help">
            To submit review please make sure to set
            <span className="reviews__star">rating</span>
            and describe your stay with at least
            <b className="reviews__text-amount">50 characters</b>.
          </p>
          <button
            className="reviews__submit form__submit button"
            type="submit"
            disabled={this.state.isFormValid && !this.state.isFormWaitingResponse
              ? `` : `disabled`}
          >
            Submit
          </button>
        </div>
        {this.state.reviewPostingError === ResponseType.ERROR && (
          <p className="reviews__error-container">
            posting review error
          </p>
        )}
      </form>
    );
  }
}

ReviewForm.propTypes = {
  offerId: PropTypes.string.isRequired,
  postReviewAction: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  postReviewAction(reviewData) {
    dispatch(postReview(reviewData));
  },
});

export {ReviewForm};
export default connect(undefined, mapDispatchToProps)(ReviewForm);
