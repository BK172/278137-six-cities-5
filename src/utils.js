import _ from "lodash";

const MarkupCitiesList = [
  `Paris`,
  `Cologne`,
  `Brussels`,
  `Amsterdam`,
  `Hamburg`,
  `Dusseldorf`,
];

export const ResponseType = {
  SUCCESS: `SUCCESS`,
  ERROR: `ERROR`,
};

export const HttpCode = {
  UNAUTHORIZED: 401,
};

export const AppRoute = {
  MAIN: `/`,
  LOGIN: `/login`,
  FAVORITES: `/favorites`,
  ROOM: `/offer/:id`,
};

export const APIRoute = {
  OFFERS: `/hotels`,
  LOGIN: `/login`,
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const getElementWidthByRating = (rating) => Math.round(rating) * 20;

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const initialSortingType = `popular`;
export const SortingTypes = {
  'popular': `Popular`,
  'to-high': `Price: low to high`,
  'to-low': `Price: high to low`,
  'top-rated': `Top rated first`
};

export const orderCitiesByList = (cities) => {
  const orderedCities = [];

  MarkupCitiesList.forEach((cityName) => {
    const city = cities.find((city) => city.name === cityName);
    city && orderedCities.push(city);
  });

  return orderedCities;
};

export const getCitiesFromOffersList = (offers) => {
  const cities = offers.map((offer) => offer.city);
  return _.uniqWith(cities, _.isEqual);
};

export const citiesAdapter = (city) => {
  return {
    name: city.name,
    coordinates: [city.location.latitude, city.location.longitude],
    zoom: city.location.zoom,
  };
};

export const offersAdapter = (offer) => {
  return {
    bedrooms: offer.bedrooms,
    city: citiesAdapter(offer.city),
    description: offer.description,
    facilities: offer.goods,
    owner: {
      id: offer.host.id,
      avatar: offer.host.avatar_url,
      name: offer.host.name,
      super: offer.host.is_pro,
    },
    offerId: offer.id,
    photos: offer.images,
    favorite: offer.is_favorite,
    premium: offer.is_premium,
    coordinates: [offer.location.latitude, offer.location.longitude],
    zoom: offer.location.zoom,
    guests: offer.max_adults,
    image: offer.preview_image,
    price: offer.price,
    rating: offer.rating,
    title: offer.title,
    type: offer.type,
  };
};
