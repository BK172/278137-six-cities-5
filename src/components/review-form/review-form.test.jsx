import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import {BrowserRouter} from "react-router-dom";
import configureStore from "redux-mock-store";
import {makeInitialStateMock} from "../../utils";
import {ReviewForm} from "./review-form";

it(`Should ReviewForm render correctly`, () => {
  const store = configureStore()(makeInitialStateMock());
  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <ReviewForm
              rating={``}
              review={``}
              isFormValid={true}
              isFormWaitingResponse={false}
              reviewPostingError={``}
              onFormSubmit={()=>{}}
              onInputChange={()=>{}}
            />
          </BrowserRouter>
        </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
