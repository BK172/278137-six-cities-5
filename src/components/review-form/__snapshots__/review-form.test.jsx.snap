// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ReviewForm render correctly 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="perfect"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="5"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="perfect"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="good"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="4"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="good"
      title="good"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="not bad"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="3"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="not bad"
      title="not bad"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="badly"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="2"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="badly"
      title="badly"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="terribly"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="1"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="terribly"
      title="terribly"
    >
      <svg
        className="form__star-image"
        height={33}
        width={37}
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    id="review"
    name="review"
    onChange={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
    value=""
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set
      <span
        className="reviews__star"
      >
        rating
      </span>
      and describe your stay with at least
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled=""
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
